### DISCUSSION TITLE
How to run unit tests against installed OpenSSL?

### DISCUSSION THREAD

**Discussion by tom-cosgrove-arm** [openssl/openssl#24925] [Q&A]
Obviously `make test` will run tests against the OpenSSL in the current build directory - it makes sense to build, test, then install. And you want to test what you've built. So the library statically linked into the tests.

But is it possible to build the tests so they dynamically link to libcrypto.so in the build directory?

**Comment by rossburton** [comment]
This query came via me: the background is that we're copying the test suite from the build tree into distro packages, so that we can run it on the target as a form of integration testing (if there's a better solution for integration testing of openssl, please say!).

The test suite is _huge_ because each test binary statically links to `libcrypto` and/or `libssl` (500M in the build tree).  This happens because there's a lot of internal symbols in the libraries that the test suite use but the shared libraries don't expose.

I'm thinking an alternative might be to build a second copy of the libraries (same objects, just a second link) that is not installed and doesn't have the linker script passed to limit symbol visibility. This will make the build faster (as it doesn't have to copy several megabytes per test binary) and smaller (one copy instead of a hundred).

**Comment by levitte** [comment]
I would say that running *the whole* test suite on installed files is quite overkill. Some of them test if internal things are working, or that the libraries built correctly. So a subset is advisable indeed. 

But either way, the current test suite isn't really designed for installations. If you get it to work, good for you!

That being said, a smaller installation test suite is a good idea, and it has been mentioned in our team, but there is yet nothing concrete that's come out of it.

**Reply by rossburton** [reply]
We basically have 100% pass rate with the test suite installed, so it's doable just annoying finding the files that the tests want.

**Reply by t8m** [reply]
IMO the only thing that could in reality get wrong between the build and deployment machine is the CPU support for the crypto algorithms. To test these, running the evp_test would be pretty sufficient. And evp_test is already dynamically linked to libcrypto so that should really test the installed library, not itself.
